### PR TITLE
feat(clustering) retiring Serf - new cache invalidation mechanism

### PR DESCRIPTION
**Pull request by thibaultcha**
This rewrite makes Kong independent when dealing with cache invalidation events. This opens the way to #2355 and to a simpler installation and usage experience for Kong.

### Incentive

As we know, Kong caches datastore entities (configuration entities mostly, such as APIs, Consumers, Plugins, Credentials...) to prevent unnecessary overhead and database round trips on a per-request basis. As of Kong 0.10, Serf has always been the dependency Kong relies upon to broadcast invalidation events to other members of the cluster. This approach is not without compromises, the main ones being:

- A dependency to a third-party software (not Nginx/OpenResty related).
- The need for a CLI to start and stop Kong, vs. the Nginx binary: because the Serf agent must be started as a daemon.
- This makes deployments and containerization somewhat awkward.
- Invoking Serf from a running Kong node requires some blocking I/O.

### New Implementation

TLDR: database polling of a time-series entity called "cluster events".

Cluster events are implemented in both PostgreSQL and Cassandra. A Kong node can `broadcast()` an event on containing data (text) on a channel, and other Kong nodes can `subscribe()` to a channel and receive said events at a configurable interval.

Here is a high-level, non-exhaustive list of changes:

- New `cluster_events.lua` module. An abstract framework for both PostgreSQL and Cassandra to broadcast messages across the cluster.
- As per our defined constant, events stay in the database for 1 day (TTL)
- For Cassandra, our table contains events with a partition key made of the channel they are published to. More details in the related commit, but this means we are quite efficient in R/W operations on this table, but that we can handle roughly 2 billion events a day. TTLs are natively supported.
- For PostgreSQL, we use a trigger on insert to handle the TTL of expired events.
- Cassandra has pagination support when reading events, PostgreSQL doesn't (yet).
- The interval at which events are polled is configurable (5 seconds by default).
- To mitigate distributed environments with eventual consistency behaviors, we introduce a configurable "propagation delay" to delay events handling until the data should have been propagated. This is an initial implementation.
- The new `mlcache.lua` module is a temporary module waiting the completion of [lua-resty-mlcache](https://github.com/thibaultcha/lua-resty-mlcache), and potentially the replacement of [lua-resty-worker-events](https://github.com/Mashape/lua-resty-worker-events).
- A new `cache.lua` module replaces the old `database_cache.lua` module. It handles inter-worker and inter-nodes invalidation, and relies on `mlcache.lua` for better performance and more powerful features.
- All entities accessed via `database_cache.get_or_set()` have been replaced.
- We try to provide node-consistency (request made on Admin API of a node instantly gets reflected in the node's proxy server) for entities. Currently, this is only done for the router (could be improved). Other entities like the balancer will need some refactoring to easily achieve this, and smaller, single entities will need an IPC module offering more consistency.
- To generate the cache key of an entity, we introduce a new schema property in most entities: `cache_key`, which can be a table taking multiple values. This makes cache key generation based on variable(s) easier and safer.
- Cached entities now have a TTL (previously was "forever") of 1 hour, and 5 minutes for a lookup miss. This is a change of default behavior as a safeguard against potentially stale data.
- As another safeguard to potentially safe data, the `/cache` endpoint on the Admin API has been kept.
- We remove [mediator_lua](https://github.com/Olivine-Labs/mediator_lua) and use lua-resty-worker-events instead. The later can achieve the job we used the former for, and has more essential features for our usage. This was a duplicate/unnecessary dependency.
- We created the skeleton for a pub/sub API to subscribe to CRUD operations on database entities. Such as `apis:updated` or `consumers:deleted`. Plugins should be able to listen on those events and act accordingly.
- With this change, Kong becomes purely **stateless**. There is no `nodes` table anymore storing each and every node currently connected to the cluster. This is by design for now, until the need to do so eventually arises.
- Core entities invalidation tests have been moved to a new test suite: `spec/04-invalidations` and are written in a more efficient manner than the previously existing tests, avoiding too many "sleep" calls (NOP) to wait for events propagation. This is thanks to the new design and will be improved with a better upcoming IPC module ensuring inter-worker consistency for each request - later on.

### Path forward

This should be reviewed and **heavily tested**. There should be some more tests written, especially around the `cluster_events` module itself. A particular attention should be given to the invalidation of core entities, and attention given there would be greatly appreciated.

Reviews and comments are welcome, questions as well! Local testing of this would be extremely helpful.

Regarding CI, the Travis builds might still fail at this point, partly for unrelated reasons (non-deterministic tests), partly because this is still in a "polishing" state.

Here is a small list of leftovers todos:

- [x] Run the new invalidation test suite on Travis CI
- [x] Find a way to test the `quiet` option of the DAO (they are currently commented out for linting reasons)
- [x] Write more tests for the cluster_events module with both databases
- [x] Audit the lua-resty-lock options of the `mlcache` module.

### CODE REVIEW

File: kong/cluster_events.lua [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+local mt = { __index = _M }
+
+
+function _M.new(opts)
+  if ngx.get_phase() ~= "init_worker" and ngx.get_phase() ~= "timer" then
+    return error("kong.cluster_events must be created during init_worker phase")
+  end
+
+  if not ngx_debug and _init then
+    return error("kong.cluster_events was already instanciated")
```
Comment (nijikokun) [inline kong/cluster_events.lua]: Reading through and caught this:

`instanciated` -> `instantiated`

File: kong/core/handler.lua [hunk 2]
```diff
@@ -31,37 +37,197 @@ local function get_now()
...
+            cache:invalidate(cache_key)
+          end
+        end
+
+        if not data.operation then
+          log(ngx.ERR, "[events] missing operation in crud subscriber")
+          return
+        end
+
+        local entity_channel           = fmt("%s", data.schema.table)
```
Comment (p0pr0ck5) [inline kong/core/handler.lua]: what is `data.schema.table` here? why do we need to wrap it in `fmt`?

File: kong/mlcache.lua [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+        return nil, "could not read from lua_shared_dict: " .. err
+    end
+
+    if v ~= nil then
+        local str_serialized, value_type, at, ttl = unmarshallers.shm_value(v)
+
+        local remaining_ttl = ttl - (now() - at)
+
+        -- value_type of 0 is a nil entry
+        if value_type == 0 then
```
Comment (p0pr0ck5) [inline kong/mlcache.lua]: i wonder if we should add an unmarshaller for nil types, so we avoid a branch here? such a dispatch could simply just return `nil` itself, so we would return the return signature of `remaining_ttl, nil, nil`. my guess is we'll have quite a lot of negative caches, so this branch becomes fairly unbiased (though this is a fairly subjective assumption).

File: kong/cluster_events.lua [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+local ngx_now   = ngx.now
+local timer_at  = ngx.timer.at
+
+
+local NODE_ID_KEY    = "cluster_events:id"
+local POLL_LOCK_KEY  = "cluster_events:poll"
+local CURRENT_AT_KEY = "cluster_events:at"
+
+
+local POLL_LOCK_EXPTIME = 60 * 5       -- 5 mins
```
Comment (p0pr0ck5) [inline kong/cluster_events.lua]: should we derive this from something configurable, such as `db_update_frequency`, rather than a hardcoded value?

File: kong/mlcache.lua [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+local type         = type
+local pcall        = pcall
+local error        = error
+local shared       = ngx.shared
+local tostring     = tostring
+local tonumber     = tonumber
+local setmetatable = setmetatable
+
+
+local LOCK_KEY_PREFIX         = "lua-resty-mlcache:lock:"
```
Comment (p0pr0ck5) [inline kong/mlcache.lua]: pedantic: do we call this key `lua-resty-mlcache` when this lib is not technically lua-resty-mlcache?

File: kong/mlcache.lua [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+        if data == CACHE_MISS_SENTINEL_LRU then
+            return unlock_and_ret(lock, nil)
+        end
+
+        return unlock_and_ret(lock, data)
+    end
+
+    -- still not in shm, we are responsible for running the callback
+
+    local ok, err = pcall(cb, ...)
```
Comment (p0pr0ck5) [inline kong/mlcache.lua]: lets say something super crappy happens here, like the callback causes the worker to crash. would we end up with a stale lock in the shm? is this a case for making the resty.lock timeout configurable, or at least thinner than the default (30 secs)?

File: kong/core/handler.lua [hunk 7]
```diff
@@ -31,37 +37,197 @@ local function get_now()
...
+        end
+
+        for i = 1, #rows do
+          local sni = rows[i]
+
+          cache:invalidate("pem_ssl_certificates:" .. sni.name)
+          cache:invalidate("parsed_ssl_certificates:" .. sni.name)
+        end
+      end, "crud", "ssl_certificates")
+
```
Comment (p0pr0ck5) [inline kong/core/handler.lua]: style: missing second whitespace line here

File: kong.conf.default [hunk 8]
```diff
@@ -328,72 +328,48 @@
...
+
+#db_cache_ttl = 3600             # Time-to-live (in seconds) of an entity from
+                                 # the datastore when cached by this node.
+                                 # If set to 0, such cached entities never
+                                 # expire.
+
+#db_cache_miss_ttl = 300         # Time-to-live (in seconds) of a miss from the
+                                 # datastore. Such misses can be cached for a
+                                 # different period of time that hits
+                                 # (configurable via `cache_ttl`).
```
Comment (Tieske) [inline kong.conf.default]: this is not necessary, as long as we consistently implement the "add" hook/event for each entity. Invalidation upon adding will remove the cache-miss-sentinel.

File: kong.conf.default [hunk 9]
```diff
@@ -328,72 +328,48 @@
...
+                                 # a multi-datacenter Cassandra cluster, this
+                                 # value should be the maximum number of
+                                 # seconds taken by Cassandra to propagate a
+                                 # row to other datacenters.
+                                 # When set, this property will increase the
+                                 # time taken by Kong to propagate the change
+                                 # of an entity.
+                                 # Single-datacenter setups or PostgreSQL
+                                 # servers should suffer no such delays, and
+                                 # this value can be safely set to 0.
```
Comment (Tieske) [inline kong.conf.default]: I do not understand what this setting does?

File: kong/cluster_events.lua [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+end
+
+
+poll_handler = function(premature, self)
+  if premature or not self.polling then
+    -- set self.polling to false to stop a polling loop
+    return
+  end
+
+  local ok, err = timer_at(self.poll_interval, poll_handler, self)
```
Comment (Tieske) [inline kong/cluster_events.lua]: a probably non-problem here might be that the `self` here will never be GC'ed. So if this `self` object might be renewed, and disposed of during the lifetime of the worker process, then there is a problem, otherwise not.

File: kong/cluster_events.lua [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+    log(ERR, "poll() threw an error: ", perr)
+
+  elseif not perr then
+    log(ERR, "failed to poll: ", err)
+  end
+
+  -- unlock
+
+  self.shm:delete(POLL_LOCK_KEY)
+end
```
Comment (Tieske) [inline kong/cluster_events.lua]: if you delete the lock, then another worker might do a poll in between, hence shortening the poll interval. Or am I missing something?

t in seconds;
t1 - worker 1 does a poll
t1.5 - worker 1 completes the poll, deletes the lock
t3 - worker 2 tries the lock, gets it and starts a poll
t3.5 - worker 2 completes the poll
t6 - worker 1 tries again (every 5 seconds)
etc.

the above will have done 2 polls in 5 seconds. Correct?

Possible way of working;
 - all workers start polling at 0.1 second intervals
 - worker starting a poll acquires lock, with a timeout of 5 seconds
 - does the poll thing, but never unlocks.
 - all other workers will not get the lock and hence not poll
 - first worker to try after the lock expired will do the next poll
etc.

File: kong/cache.lua [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+local type    = type
+local max     = math.max
+local ngx_log = ngx.log
+local ngx_now = ngx.now
+local ERR     = ngx.ERR
+local NOTICE  = ngx.NOTICE
+local DEBUG   = ngx.DEBUG
+
+
+local SHM_CACHE = "cache"
```
Comment (thibaultcha) [inline kong/cache.lua]: This is not a namespace prefix, but the name of the shm entry.

File: kong/cmd/restart.lua [hunk 13]
```diff
@@ -10,6 +10,7 @@ local function execute(args)
   end
 
   pcall(stop.execute, args)
+  ngx.sleep(0.1)
```
Comment (thibaultcha) [inline kong/cmd/restart.lua]: I have to do a small update here, ignore.

File: kong/cache.lua (hunk 12 above)
Comment (p0pr0ck5) [inline kong/cache.lua]: yes, "namespace" is colloquial, referring to usage within the shm. it may be possible that other users want to use this shm? we should be a good citizen imo and be clear about what our keys are doing in the shm.

File: kong/plugins/rate-limiting/policies/init.lua [hunk 14]
```diff
@@ -28,11 +28,18 @@ return {
       for period, period_date in pairs(periods) do
         if limits[period] then
           local cache_key = get_local_key(api_id, identifier, period_date, period)
-          cache.sh_add(cache_key, 0, EXPIRATIONS[period])
 
-          local _, err = cache.sh_incr(cache_key, value)
-          if err then
-            ngx_log("[rate-limiting] could not increment counter for period '"..period.."': "..tostring(err))
+          local ok, err = shm:add(cache_key, 0, EXPIRATIONS[period])
```
Comment (thibaultcha) [inline kong/plugins/rate-limiting/policies/init.lua]: I remember this issue being raised already. We could update this in this PR, or in another one :)

File: kong/mlcache.lua (hunk 6 above)
Comment (thibaultcha) [inline kong/mlcache.lua]: This is why this PR says we need to audit those settings, another small adjustment to make.

File: kong/mlcache.lua (hunk 3 above)
Comment (thibaultcha) [inline kong/mlcache.lua]: I would not address this here, but in the actual lua-resty-mlcache repo

File: kong.conf.default (hunk 8 above)
Comment (thibaultcha) [inline kong.conf.default]: Both of those TTL values are a safeguard against eventually stale data. It's a very bad idea imo to not have a TTL in our cache, and to solely rely on the invalidation events being called. It could go wrong, somehow...

File: kong/cluster_events.lua (hunk 10 above)
Comment (thibaultcha) [inline kong/cluster_events.lua]: Yeah it should be fine, because we ensure that instances of this module are singletons.

File: kong.conf.default (hunk 8 above)
Comment (Tieske) [inline kong.conf.default]: yes, but we don't need another one for the cache-miss, just use the regular one, as it is only a safeguard.

Comment (thibaultcha) [inline kong.conf.default]: I'd be fine with not having it. It's just that they offer better granularity and are more future-proof, with minimum cognitive overhead from a user friendliness perspective...

**Comment by thefosk** [comment]
Turns out that we may want to keep Kong nodes registering to the database, and having node health-checks, to show this information in our other products.

**Comment by thibaultcha** [comment]
I know, this can come in another PR though, with more meaningful data and a better and more popularized implementation.

File: kong.conf.default (hunk 8 above)
Comment (Tieske) [inline kong.conf.default]: thinking about this again; shouldn't there be a setting that determines how far back events are being dealt with, than the whole expiry thing is not necessary right? if a case occurs where events are missed, then increasing that setting should suffice. If not, there is a serious bug somewhere. Correct?

File: kong/cluster_events.lua (hunk 11 above)
Comment (p0pr0ck5) [inline kong/cluster_events.lua]: An alternative approach could be to leverage the `flags` portion with a number representing `ngx.now()` + `poll_interval`, and the lock can only be acquired when the poller's `ngx.now()` is greater than such a value. But this becomes problematic with poll intervals < 1s, and feels kinda fishy.

Alt-alternatively, we can solve the problem of misaligned workers as @Tieske described by forcing each execution to a floor period of time modulo poll interval, e.g., with an interval of 5s, each worker is forced to run at t1, t6, t11, etc... by calculating the timer_at as a diff from `now()` and the next floor up. Of course, both of these scenarios imply that we care about explicitly and consistently running polls exactly at every `poll_interval` seconds- do we?

Comment (thibaultcha) [inline kong/cluster_events.lua]: @Tieske The behavior you described was the one in place before a short review by @p0pr0ck5 a few weeks ago, in a private fork. From the top of my head, I forgot what was the original, other issue, but will try to find it, and if none, then yes, we can come back to this behavior which also has the benefit of having the smaller cognitive load.

File: kong.conf.default (hunk 8 above)
Comment (thibaultcha) [inline kong.conf.default]: Removed this setting to rely on `db_cache_ttl` in the latest version of this branch.

File: kong/core/handler.lua (hunk 2 above)
Comment (thibaultcha) [inline kong/core/handler.lua]: Historically I had different channels, then updated them, but did not remove the `fmt` call. We can for sure :)

File: kong/core/handler.lua [hunk 15]
```diff
@@ -31,37 +37,198 @@ local function get_now()
...
+        end
+
+        if not data.operation then
+          log(ngx.ERR, "[events] missing operation in crud subscriber")
+          return
+        end
+
+        local entity_channel           = fmt("%s", data.schema.table)
+        local entity_operation_channel = fmt("%s:%s", data.schema.table,
+                                             data.schema.operation)
```
Comment (thibaultcha) [inline kong/core/handler.lua]: Big mistake. This should be `data.operation`. This is one of the things we'll need tests for once the design is ok for everybody.

File: kong/cluster_events.lua (hunk 11 above)
Comment (thibaultcha) [inline kong/cluster_events.lua]: Ok, asked @p0pr0ck5 to remind me: it was that if a poll takes longer to execute than the lock exptime, we'd have 2 running timers with poll at the same time.

Here is a 3rd alternative: use 2 locks, one with exptime, one removed when the poll is complete by the current worker, and only allow a worker to run the lock if both lock are available.

Comment (Tieske) [inline kong/cluster_events.lua]: > it was that if a poll takes longer to execute than the lock exptime, we'd have 2 running timers with poll at the same time.

Yes, that is the drawback indeed.

> use 2 locks, one with exptime, one removed when the poll is complete by the current worker, and only allow a worker to run the lock if both lock are available.

I like this one best. As that would be the more elegant solution.

Now what are sensible defaults then? lock 1 has a 5 seconds timeout (causing a 5 second interval). Lock 2 has a 30 second (??) timeout as a safety feature? and only being able to acquire 1 lock issues a warning in the logs right?

How often will each worker try to acquire the locks to do a poll? 1/10th of the poll interval, eg. 5 seconds => 0.5 second. Then the poll interval will be between 5 and 5.5 seconds. With 4 workers the average interval time would then be 0.5/4 + 5 = 5.125 seconds

**Comment by mwaaas** [comment]
Is there a eta for this "With this change, Kong becomes purely stateless. " currently blocking us from using it in eb docker

**Comment by p0pr0ck5** [comment]
@Tieske can you help me clarify your meaning behind this comment:

> eg. grabbing an api from an lru cache that also holds 10000 consumers takes longer than necessary if we would split them up.

LRU cache operations are O(1). There is no performance gain by using multiple LRU caches (if anything, it's just more complexity in management).

File: kong/cluster_events.lua (hunk 11 above)
Comment (thibaultcha) [inline kong/cluster_events.lua]: Switched to use a double lock mechanism now

File: kong/mlcache.lua (hunk 6 above)
Comment (thibaultcha) [inline kong/mlcache.lua]: This is why there was a timeout value in the original implementation. In the current one with a double lock mechanism. there is a timeout for both locks as well, in case such a scenario happens.

File: kong/mlcache.lua [hunk 16]
```diff
@@ -0,0 +1,466 @@
...
+
+    self.ipc = ipc
+
+    return setmetatable(self, mt)
+end
+
+
+function _M:purge()
+    self.dict:flush_all()
+    self.dict:flush_expired()
```
Comment (thibaultcha) [inline kong/mlcache.lua]: A boolean value could be implemented here, that'd be neat, and yes, using the LRU mechanism is definitely fine I think

File: kong/cmd/restart.lua (hunk 13 above)
Comment (thibaultcha) [inline kong/cmd/restart.lua]: Now properly implemented with a `kill -0` loop to make sure Nginx is properly terminated before trying to start it again.

File: kong.conf.default (hunk 9 above)
Comment (thibaultcha) [inline kong.conf.default]: Updated the description to clarify the meaning

File: kong/plugins/acl/hooks.lua [hunk 17]
```diff
@@ -1,26 +0,0 @@
-local events = require "kong.core.events"
```
Comment (thibaultcha) [inline kong/plugins/acl/hooks.lua]: We now handle all invalidation events that are easily deducible (when the entity being invalidated is the same as the one being updated, or when it is a foreign entity that we can track because of our schema's `foreign` property) via a new property on the schema, which is `cache_key`. When an entity, or a parent entity is being updated, we compute its cache_key via its schema, and tell all the other Kong nodes "that's the unique cache key you need to invalidate".

File: kong/mlcache.lua [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+        end
+
+        return json
+    end,
+}
+
+
+local unmarshallers = {
+    shm_value = function(marshalled)
+        local ttl_last = find(marshalled, ":", 21, true) - 1
```
Comment (p0pr0ck5) [inline kong/mlcache.lua]: @thibaultcha ping ^ this would be nice to have

Comment (thibaultcha) [inline kong/mlcache.lua]: Added some

File: kong/mlcache.lua (hunk 5 above)
Comment (thibaultcha) [inline kong/mlcache.lua]: Yeah, I think that's fine, maybe we'll get to replacing this file with the upstream mlcache lib

File: kong/cache.lua (hunk 12 above)
Comment (thibaultcha) [inline kong/cache.lua]: Alright, let's take advantage of the fact that 0.11 will have many updates to the Nginx config template already to chose a better name, and group the breaking changes in one.

I suggest we prefix all of our shms with `kong_*`. How about: `kong_db_cache` for this one?

Comment (p0pr0ck5) [inline kong/cache.lua]: sgtm!

Comment (thibaultcha) [inline kong/cache.lua]: Cool. Let's do that right after the merge, to have one single commit taking care of them all at once, to preserve atomicity then.

**Comment by olderwei** [comment]
since the version 0.11 has cancelled the command of kong cluster, how do I know which nodes make up a cluster

**Comment by thibaultcha** [comment]
@olderwei In 0.11 you cannot: all Kong nodes are purely stateless. The `kong cluster` command was only useful for debugging purposes when a Kong node's Serf agent seemed to not be receiving events. You should use orchestration tools to manage your cluster and have visibility into its current state.

**Comment by olderwei** [comment]
@thibaultcha  so do you have any recommendations about the orchestration tools?
